Adding spider for Sobral/CE

camilafracaro opened the pull request.

camilafracaro commented:
Hello, I think it's ready for review. I tried to merge my 2 commits into 1, but I just couldn't :(
I tested and it was downloading from end of May to now and I also tested not saving the files. It is running one month at a time.

jaswdr reviewed (changes requested):
Hello @camilafracaro, thanks a lot for your PR, I was checking the code and see some opportunities of improvements, please take a look in this refactor that I made https://gist.github.com/jaswdr/14c1ad381ae56d71b104f73e4a4b3c7c

In general, what I change was this:
1. Remove the usage of `is not None` in your if statements, you don't need it, using just `if(something)` is the same;
1. Keep the variable's instantiation near the usage, this way the person that is reading don't need to scroll up and down to understand what is happening;
1. Removed the `done` variable, you don't need it, the checks that you already are doing with your `if` statements are enough;

Otherwise, the code looks good to me, I liked that you are checking for the next page in the end and starting a new request in a recursive way, good job.

Feel free to take a look at what I have put above, and the file refactored that I shared.

camilafracaro commented on data_collection/gazette/spiders/ba_salvador.py:
```diff
@@ -47,7 +47,9 @@ def parse(self, response):
     def parse_gazette(self, response, gazette_date):
         gazette_date = datetime.datetime.strptime(gazette_date, "%Y-%m-%d").date()
-        pdf_url = response.css("#PDFId embed::attr(src)").get()
+        pdf_url = response.css("#PDFId object::attr(data)").get()
```
This "change" is in production already, I didn't change it on mine, but I did something weird when I was trying to push it from upstream. I do remember having problems with it :)

camilafracaro commented:
Hi @jaswdr, 
thanks for reviewing it and for the tips - I'm still learning Python and I appreciate it a lot!
The new commit is basically your code, but I did 4 changes:
- Added https because it was redirecting from http to https - I think this is new to the site, because I didn't have this before.
- Line 30: total gazettes always have a value, so put to return if == 0 instead of not total_gazettes - it's only in a few cases like accessing on the 1st of the month when there's no gazettes.
- Line 58: added an if clause because if there's only one page (beginning of the month) we don't have the links with the pages and next_page will fail if current_page is none. 
- Based on your second point, I reordered some lines to be close to their usage.

About the other files that were not supposed to be on my commits, is there anything I can do about it? I always struggle to update with main project before sending the commits.

Let me know if I need to change anything else and thanks again!
Ps.: I hope pubs are finally totally open next week and you can enjoy the cold Irish summer :)

jaswdr commented:
> Line 30: total gazettes always have a value, so put to return if == 0 instead of not total_gazettes - it's only in a few cases like accessing on the 1st of the month when there's no gazettes.

Is not only for not having a value, the `if not total_gazettes` handles cases where `total_gazettes` is 0, or if it is an empty list, or if is None, or if is False, or any other value that is translated to `False`, this is a common pattern in Python and I would recommend keeping it as `if not total_gazettes` to have a better coverage.

Anything else, LGTM

camilafracaro commented:
Hi @jaswdr,
I just tested the not total_gazettes and because it comes as a text (" 0"), it won't be true. Do you think it's better to keep "not" because the page can change? In this case, isn't it better to fail to be easier to be noted? Otherwise we will continue to run the spider with no error as it is going to return.

My test is below. 
>>> from lxml import html
>>> import requests
>>> page = requests.get('https://www.sobral.ce.gov.br/diario/pesquisa/index/ano_da_publicacao:2021/mes_da_publicacao:08')
>>> tree = html.fromstring(page.content)
>>> trs = tree.xpath("//div[@class = 'right']/text()")
>>> trs
[' 0']
>>> not trs[0]
False
>>> not '0'
False
>>> not 0
True

jaswdr commented:
> Hi @jaswdr,
> 
> I just tested the not total_gazettes and because it comes as a text (" 0"), it won't be true. Do you think it's better to keep "not" because the page can change? In this case, isn't it better to fail to be easier to be noted? Otherwise we will continue to run the spider with no error as it is going to return.
> 
> 
> 
> My test is below. 
> 
> >>> from lxml import html
> 
> >>> import requests
> 
> >>> page = requests.get('https://www.sobral.ce.gov.br/diario/pesquisa/index/ano_da_publicacao:2021/mes_da_publicacao:08')
> 
> >>> tree = html.fromstring(page.content)
> 
> >>> trs = tree.xpath("//div[@class = 'right']/text()")
> 
> >>> trs
> 
> [' 0']
> 
> >>> not trs[0]
> 
> False
> 
> >>> not '0'
> 
> False
> 
> >>> not 0
> 
> True

Good catch, I agree, considering that the value ' 0' is possible, you are right, please keep your solution.
